elliptic-curve/src/hash2curve/group_digest.rs

```diff
@@ -1,17 +1,20 @@
 use super::MapToCurve;
 use crate::{
     hash2field::{hash_to_field, ExpandMsg, FromOkm},
-    Result,
+    ProjectiveArithmetic, Result,
 };
 use group::cofactor::CofactorGroup;
 
 /// Adds hashing arbitrary byte sequences to a valid group element
-pub trait GroupDigest {
+pub trait GroupDigest: ProjectiveArithmetic<ProjectivePoint = Self::Output> {
     /// The field element representation for a group value with multiple elements
-    type FieldElement: FromOkm + Default + Copy;
+    type FieldElement: FromOkm
+        + MapToCurve<Output = Self::Output>
+        + Default
+        + Copy
+        + Into<Self::Scalar>;
```

Comment:
I added this to implement `hash_to_scalar`, unsure if this is really what we need here. My assumption is that `FieldElement` to `Scalar` conversion doesn't need modulo or anything special, is that wrong?

Reply:
`FieldElement` is an element of the base field.

`Scalar` is an element of the scalar field.

They're two completely different fields with two different moduli.

Outside of an extremely limited set of legacy algorithms (i.e. ECDSA) you shouldn't be converting an element of one field into the other.

Reply:
Consumers shouldn't be operating on `FieldElement` directly as they do with `Scalar`. I'm not even sure why you would want to expose `FieldElement`. I used to need it to implement hash_to_curve but now that's taken care of.

Reply:
I would definitely prefer `FieldElement` remain hidden from the public API, because it's extremely easy to misuse (especially `k256` with lazy normalization).

If things legitimately need direct access to `FieldElement`, I'd prefer they get added directly to `k256`/`p256`.

Reply:
I don't want to expose it, I want to somehow use `hash_to_field` an return a Scalar.

> `FieldElement` is an element of the base field.
> 
> `Scalar` is an element of the scalar field.
> 
> They're two completely different fields with two different moduli.

Clearly I don't know what I'm doing, if we can't convert a `FieldElement` to a `Scalar`, what's the way to make `hash_to_scalar` work?

Reply:
I reverted the whole thing for now.

Reply:
`hash_to_field` seems to be defined for any field, so you perform the same set of operations you would with `FieldElement`, but use operations on `Scalar` instead.

Sidebar: it's really annoying they defined a complicated algorithm for this rather than just use a wide modular reduction. I see Ristretto255 does the latter. Kind of confused why they didn't do the same for the NIST P-curves but oh well.

Reply:
Yeah, I didn't realize that, so what I'm currently missing is how to do the `FromOkm`, the way that @mikelodder7 did it is something I don't understand, I simply did modulo reduction, which seemed to work. I will look into it further.